CAEP for VMs as bastion hosts (#1496)
kubernetes-sigs/cluster-api-provider-azure: docs/proposals/20210705-bastion-vms.md

## Proposal

To adhere to the KISS principle the idea is to reuse existing reconciliation controllers for creating the `Bastion VM(s)`.
In practice that means creating 3 resources:

1. A `Secret` holding the `cloud-init` config file to bootstrap the bastion VM with the needed config (has to be provided by user and can't be defaulted in any way)
2. An `AzureMachineTemplate` for each bastion that describes how the VM should be (size, OS, disks, etc). Can be defaulted.
3. A `MachineDeployment` to bind the two aformentioned resources and make CAPI/CAPZ controllers reconcile and create the VMs. Can be defaulted.

The suggestion is to add a new subfield to the `AzureCluster/spec/azureBastion` field like this:

```yaml
apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
kind: AzureCluster
metadata:
  name: cluster-name
  namespace: default
spec:
  azureBastion:
    bastionHosts:
    - bootstrapDataSecretName: "<secret name>"
      azureMachineTemplate: {}
      replicas: 2
```

With the above information, the `AzureCluster controller` can create an `AzureMachineTemplate` and a `MachineDeployment` and then the creation of VMs should happen automagically.

> Why a slice of `bastionHosts` rather than a single value? Because having multiple bastion hosts only makes sense if they are placed in different availability zones.

Reviewer: What do you think about making this a Virtual Machine Scale Set rather than individual machines. On a VMSS, you can specify the zones you would like and the platform will balance the machines. Also, if you update the cloud-init, you could have the VMSS automatically roll out the change.

Author: I am not sure I like it.
In general I'm all in for VMSSes in place of VMs, but the main point of having more than one bastion is to distribute them in different availability zones.
VMSS have multi AZ support, but customers don't have control in which AZs VMSS instances are run into.
Potentially all instances can end up in the same AZ and, in general, it is impossible to force AZ allocation.
This is why I went down the VM road.
WDYT?

Reviewer: I believe we can achieve strict zone balancing and also enforce max spread of machines via https://docs.microsoft.com/en-us/azure/virtual-machine-scale-sets/virtual-machine-scale-sets-use-availability-zones#zone-balancing so that machines are required to be in difference AZs and different fault domains.

Author: Didn't know about this. Then it totally makes sense to switch to VMSS to me.
Will work on that. Thank you David

Author: I gave some more though time to this VMSS vs VM subject.
I started thinking that this whole CAEP might be useless.
Using `(Azure)MachinePools` to bootstrap bastion hosts requires the user to provide a whole `MachinePoolSpec` CR definition inside the `AzureCluster`. What is the point of adding it there? The user can simply create the `MachinePool` CR directly.
By the way I tried to do so, and it is currently impossible to create VMSSes as bastions using existing controllers, because:

- there is no support for adding public IPs to VMSS instances
- if you add support for public IPs bootstrap fails because of this https://github.com/Azure/aks-engine/issues/3521

I will try to work on his a little bit longer (trying to deploy VMs/VMSSes as bastions to get the feeling how it could work) but I feel like this CAEP is destined to be closed.

Reviewer 2: > Using (Azure)MachinePools to bootstrap bastion hosts requires the user to provide a whole MachinePoolSpec CR definition inside the AzureCluster. What is the point of adding it there? The user can simply create the MachinePool CR directly.

The key difference is that the bastion VM/VMSS instance shouldn't itself be a k8s node. It shouldn't have azure.json on it, and it shouldn't run kubeadm.

> if you add support for public IPs bootstrap fails because of this Azure/aks-engine#3521

I don't think that applies to CAPZ, this is just a shortcoming of AKS Engine

Reviewer: Perhaps, adding a public IP for each VMSS instance is not needed. I think this could be accomplished by adding a NAT rule to the public LB to NAT traffic to individual instances.

Since the `failureDomain` field is a scalar in the `AzureMachineTemplate` CR, it's needed to have different templates for different availability zones.

![Figure 1](./images/bastion-vms/fig1.png "Architecture of the proposal")

### User Stories

#### Story 1

Frank is a Platform Engineer for a company offering Kubernetes as a Service.
The company has a lot of kubernetes clusters and Frank, together with his colleagues, needs to access the machines
forming the Kubernetes cluster via SSH in order to provide support to his company's customers.

For security policy compliance reasons, the master node(s) of the clusters can't expose the SSH port to the public internet.

Frank writes a `MutatingWebhook` that ensures every `AzureCluster` CR contains a `spec.bastionSpec.bastionHosts` to set up 
one bastion host for every cluster.

> This way he ensures all Clusters have the same setup in terms of Bastions and he can create a tool to easily SSH into

Reviewer: Are all of the SSH keys in the cloud-init?

Author: Potentially, but this is user choice.
For example, we (Giant Swarm) would not use ssh keys but use Hashicorp's Vault for SSH purposes.
The whole point of asking the client to give a full cloud init secret is to have max flexibility for virtually any use case

Reviewer: I see you point about having flexibility.

How do you envision the behavior when an operator will mutate or replace the cloud-init secret?

Author: The bastion hosts should be rolled in my opinion.
In general, I would stick to what CAPZ is doing with node pool nodes in this case.

      * [Existing APIs for Clarity](#existing-apis-for-clarity)
      * [Proposed API Changes](#proposed-api-changes)
  * [Conclusions](#conclusions)
  * [Additional Details](#additional-details)
    * [Test Plan](#test-plan)
  * [Implementation History](#implementation-history)

## Glossary

* CAPZ - An abbreviation of Cluster API Provider Azure.
* KISS - Keep It Simple, Stupid (see [Wikipedia](https://en.wikipedia.org/wiki/KISS_principle))

## Summary

Accessing virtual machines forming a Kubernetes cluster for administrative purposes is a potential security threat.
A common pattern to provide SSH/RDP access to virtual machines without exposing them to the public internet is to use 
a `Bastion host` or `Jump host`, a dedicated machine that is exposed to the internet acting as a bridge between the administrators'
computers and the Kubernetes nodes (that can therefore stay behind a firewall).
This CAEP aims at providing a way to bootstrap `Bastion Hosts` in a CAPZ cluster levereging existing Controllers as much as possible
by reusing CRDs like `MachineDeployment` and `AzureMachineTemplate`.

## Motivation

> By default CAPZ clusters expose one of the Master nodes to the internet making it possible to access all the other nodes in the cluster via SSH or RDP.

Reviewer 2: ```suggestion
By default CAPZ clusters expose one of the control plane nodes to the internet making it possible to access all the other nodes in the cluster via SSH or RDP.
```

As an alternative, it is possible to use the [`AzureBastion`](https://azure.microsoft.com/en-us/services/azure-bastion/) feature to SSH/RDP to nodes using a web interface through 
the `Azure Portal` (see [documentation](https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/master/docs/book/src/topics/ssh-access.md)).

Those solutions might not cover all use cases. As a matter of example:

> - Using one master node as a jump host might be not compliant to the security policies for some companies.

Reviewer 2: ```suggestion
- Using one control plane node as a jump host might be not compliant with the security policies for some companies.
```

> - Using `AzureBastion` might be inpractical or unusable in some use cases (machine to machine access would be nearly impossible).

Reviewer 2: > machine to machine access would be nearly impossible

can you clarify what this means?

Author: Added a hopefully clearer explanation

### Goals

> - Allow users to deploy virtual machines as bastion hosts in multiple availability zones using a custom cloud init file.

Reviewer 2: > custom cloud init file

does this imply the bastion VMs can be Linux only? should windows bastions / other OSes like Flatcar be added as a non-goal?

Author: No at all. Allowing users to specify the config secret directly grants maximum flexibility.
For example I've been able to test this using `flatcar linux` and `ignition` so I assume Windows will be perfectly supported.

> ### Non-Goals/Future Work

Reviewer 2: should we also add removing the existing AzureBastion service support as a non-goal?

Author: I'd say no. In my vision `AzureBastion` has a different use case and it can be an alternative solution to whom that don't need/want a dedicated VM as bastion.

> - Make it possible to disable the default behaviour of exposing one master node to the internet.

Reviewer 2: ```suggestion
- Make it possible to disable the default behaviour of exposing one control plane node to the internet.
```

## Proposal

To adhere to the KISS principle the idea is to reuse existing reconciliation controllers for creating the `Bastion VM(s)`.
In practice that means creating 3 resources:

1. A `Secret` holding the `cloud-init` config file to bootstrap the bastion VM with the needed config (has to be provided by user and can't be defaulted in any way)
2. An `AzureMachineTemplate` for each bastion that describes how the VM should be (size, OS, disks, etc). Can be defaulted.
3. A `MachineDeployment` to bind the two aformentioned resources and make CAPI/CAPZ controllers reconcile and create the VMs. Can be defaulted.

The suggestion is to add a new subfield to the `AzureCluster/spec/azureBastion` field like this:

```yaml
apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
kind: AzureCluster
metadata:
  name: cluster-name
  namespace: default
spec:
  azureBastion:
    bastionHosts:

>     - bootstrapDataSecretName: "<secret name>"

Reviewer 2: this could be confused with the field `DataSecretName` same name in CABPK

Author: That's exactly what it is. That field is meant to be a reference to a secret that is identical to the secret that CABPK would create for a "traditional" Machine.
